### DISCUSSION TITLE
Run Routing Code

### DISCUSSION THREAD

**Discussion by dajinchu** [vikejs/vike#986] [Discussion]
Is there a way to run just VPS routing and get the _pageId, but not render anything? For my use case, in my fastify server I would like to check "Does VPS know how to handle this route" to decide whether or not to proxy the request to the legacy server or not. For now I am just calling renderPage but it is not an efficient solution.

[My use case](https://github.com/Alignable/bifrost/blob/05d7d9e442365694d0bd2cc8e832a5bedc138558/bifrost-fastify/index.ts#L75)

**Comment by brillout** [comment]
> For now I am just calling renderPage but it is not an efficient solution.

Your approach seems fine to me, what makes you think it isn't effecient? When VPS returns a 404 then only your Route Functions have been called and no other hook is called. So it should be fast (as long as our Route Functions are fast).

**Reply by dajinchu** [reply]
For me, when no route matches, it renders the error page which is a bit of a waste.

**Reply by brillout** [reply]
One thing you can do is to throw a special error upon a 404 in your `onBeforeRender()` hook or, if you don't have one, in your `render()`/`onRenderHtml()` hook.

Then check whether `pageContext.errorWhileRendering` matches that special error.

**Reply by dajinchu** [reply]
Ah, ok! How do I differentiate then between a 404 that is caused by no route match, vs. a 404 from an onBeforeRender throwing an `RenderErrorPage`? For us, no route match should proxy, but RenderErrorPage should actually render the 404 page. Just check errorWhileRendering inside the onRender()?

**Reply by brillout** [reply]
> Just check errorWhileRendering 

Yes I believe it should do the trick. Keep me updated how it goes.
